sim-lib/src/random_activity.rs

```diff
@@ -0,0 +1,211 @@
+/// node's capacity, it will produce payments such that the node sends multiplier * capacity over a calendar month.
+/// While the expected amount to be sent in a month and the mean payment amount are set, the generator will introduce
+/// randomness both in the time between events and the variance of payment amounts sent to mimic more realistic
+/// payment flows.
+pub struct PaymentActivityGenerator {
+    multiplier: f64,
+    expected_payment_amt: u64,
+    source_capacity: u64,
+    event_dist: Exp<f64>,
+}
+
+impl PaymentActivityGenerator {
+    /// Creates a new activity generator for a node, returning an error if the node has insufficient capacity deployed
+    /// for the expected payment amount provided. Capacity is defined as the sum of the channels that the node has
+    /// open in the network divided by two (so that capacity is not double counted with channel counterparties).
+    pub fn new(
+        source_capacity_msat: u64,
+        expected_payment_amt: u64,
+        multiplier: f64,
+    ) -> Result<Self, SimulationError> {
+        PaymentActivityGenerator::validate_capacity(source_capacity_msat, expected_payment_amt)?;
+
+        // Lamda for the exponential distribution that we'll use to randomly time events is equal to the number of
+        // events that we expect to see within our set period.
+        let lamda = events_per_month(source_capacity_msat, multiplier, expected_payment_amt)
+            / (SECONDS_PER_MONTH as f64);
+
+        let event_dist =
+            Exp::new(lamda).map_err(|e| SimulationError::RandomActivityError(e.to_string()))?;
+
+        Ok(PaymentActivityGenerator {
+            multiplier,
+            expected_payment_amt,
+            source_capacity: source_capacity_msat,
+            event_dist,
+        })
+    }
+
+    /// Validates that the generator will be able to generate payment amounts based on the node's capacity and the
+    /// simulation's expected payment amount.
+    pub fn validate_capacity(
+        node_capacity_msat: u64,
+        expected_payment_amt: u64,
+    ) -> Result<(), SimulationError> {
+        // We will not be able to generate payments if the variance of sigma squared for our log normal distribution
+        // is > 0 (because we have to take a square root).
+        //
+        // Sigma squared is calculated as: 2*ln(payment_limit) - ln(expected_payment_amt)
+        // Where: payment_limit = node_capacity_msat /2.
+        //
+        // Therefore we can only process payments if: 2ln(payment_limit) - ln(expected_payment_amt) >= 0
+        //   ln(payment_limit)      >= ln(expected_payment_amt)/2
+        //   e^ln(payment_limit)    >= e^(ln(expected_payment_amt)/2)
+        //   payment_limit          >= sqrt(expected_payment_amt)
+        //   node_capacity_msat / 2 >= sqrt(expected_payment_amt)
+        //   node_capacity_msat     >= 2 * sqrt(expected_payment_amt)
```

Comment:
Here what you really want to make sure of is that the `node_capacity_msat/2 >= expected_payment_amt`, since the `payment_limit = min(node1_cap, node2_cap)`, hence if any of the nodes has `node_capacity_msat < expected_payment_amt`, then `ln(payment_limit) < ln(expected_payment_amt)` and, therefore, `sigma` will yield `NaN`.